pep-9999.rst

```diff
@@ -0,0 +1,213 @@
+
+- Because variance is a property of the generic type itself, type 
+  checkers do no allow using co- or contravariant type variables 
+  outside of generic classes. But as the variance is currently a
+  property of the type variables, the must be used within variant
+  generic classes.
+
+- Creating separate type variables for different variances often
+  results in up to 3x code duplication, especially for type variables
+  with constraints or an upper bound. 
+
+
+Proposal
+========
+
+
+All of these issues can be solved by overloading the ``+`` and ``-``
+prefix operators of ``typing.TypeVar`` instances, so that ``+T`` is an
+alias for a covariant type variable, and ``-T`` for a contravariant 
+one. These variance operators should only be used within the parameter
+list of `typing.Generic` and `typing.Protocol` when subclassing them.
+
+This syntax is practically identical to type parameter variance in the
+Scala programming language [1]_.
+
+
+
+Specification
+=============
+
+The new type variable variance syntax should be used within the generic
+parameter list of `typing.Generic` or `typing.Protocol` base classes,
+and only there. It is not allowed to be used on type variables that
+are already marked as covariant or contravariant, either with 
+`covariant=True`, `contravariant=True` or with the new `+T` or `-T`
+syntax.
+
+Simplified Syntax
+-----------------
+
+Instead of specifying the variance on the typevar itself::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the `+` and `-` prefix operators to specify
+covariant and contravariant generic types inline::
+
+    T = typing.TypeVar('T')
+
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+
+Differences with current syntax
+-------------------------------
+
+The new typevar operators return a transparent wrapper around the 
+original type variable, which can be accessed with the `__orig__` 
+attribute on the returned wrapper. e.g.::
+
+    (+T).__orig__ is T
+    (+T).__covariant__ == not T.__covariant
+    (+T).__contravariant__ is False
+    (+T).__name__ == T.__name__
+    (+T).__constraints__ == T.__constraints__
+    (+T).__bound__ is T.__bound__
+
+
+Thus, type variables defined with `covariant=True` and 
+`contravariant=True`, are not equivalent to `+T` and `-T`.
+
+
+`+T` and `-T` are not valid type annotations, and should only be used
+within the generic type parameter list of generic classes. e.g.:: 
+
+    class Spam(typing.Generic[+KT]): ...
+    class Eggs(typing.Protocol[KT, +VT]): ...
+    class OrderedSet(typing.AbstractSet[+T]): ...
```

Comment:
This is invalid according to the specification, which says you can only put it in Generic or Protocol. I was actually going to ask whether this should be allowed.

Reply:
Ah that's good to know. I can see how relaxing that restriction can be desirable, but I guess that that's out of scope here.

Reply:
Is it a relaxing? This syntax is definitely legal right now without the `+`. If you allow it with your new syntax, you'll have to decide what it means. For example, maybe it should be possible to have an invariant subclass of a covariant class.

Reply:
Actually, thinking about it a little more I believe you should allow this, because invariant subclasses of covariant classes are commonplace. My suggestion would be that you have to specify the `+` again for every subclass, and if there's no `+` the subclass is invariant.

For example, we have this in the standard library:
```
class AbstractSet(Generic[+T]): ...  # covariant

class set(Sequence[T]): ...  # invariant

class frozenset(Sequence[+T]): ... # covariant
```
To express that, we need what I suggested above.

Reply:
I also agree with Jelle here, I think it should be allowed anywhere the the old forms are allowed.
On another note typing.AbstractSet is deprecated in favour of collections.abc.Set, so it might be a good idea to change that.

Reply:
Another edge case is this:
```
class WhatIsMyVariance(Generic[+T], Sequence[-T]): ...
```
That should probably be disallowed, but this should be spelled out.

Reply:
I agree with your suggestion @JelleZijlstra . But I think that In your example `set` and `frozenset` also need to subclass `Generic`, i.e.

```python

class set(Sequence[T], Generic[T]): ...  # invariant

class frozenset(Sequence[T], Generic[+T]): ... # covariant
```

Then the restriction that `+T` and `-T` can only be used within the type parameter list of `Generic` and `Protocol` still holds. If you were to instead write `class frozenset(Sequence[+T], Generic[+T]): ...`, to me it looks as if `Sequence` itself is marked as covariant under `T`, whereas the covariance only applies to the specific definition of `frozenset`; the variance information does not need to be passed to the `Sequence` superclass. 
This also avoids the possibility for edge cases, like in `WhatIsMyVariance`.

Reply:
That's not how things currently work, though: writing `class set(Sequence[T]):` is legal and accepted by typecheckers.

Reply:
Ah I see, somehow I thought that in that case you couldn't do e.g. `spam: set[str]`, and that it required an explicit `Generic` as additional superclass. 

Anyway, I now indeed see how this behavior has to be explicitly defined. Requiring the same variance for the same type variable in the same generic class should do the trick. So no mixing; only allow e.g.
- `class LinkedList(Iterable[T], Container[T]): ... `
- `class LinkedList(Iterable[+T], Container[+T]): ... ` 
- `class LinkedList(Iterable[-T], Container[-T]): ... ` 

This could be checked within `typing.Generic.__init_subclass__`.